### PR TITLE
Adding support for running MacOS Universal binaries

### PR DESCRIPTION
**Pull request by plk**
This patch only alters the `boot.c` code for MacOS. It:

- Checks to see if the exe is a universal binary
- If it isn't, proceed as normal
- If it is, extract the local arch version of the binary to a thin, non-universal binary in a dedicated cache dir with "_FAT" postfix
- exevce over the existing process to run the thin arch-specific binary

I had to post-process the tmpdir used as `par_mktmpdir` results in paths with repeated slashes in it which `execve` doesn't like.

One thing to check is whether generic MacOS contains `/usr/binb/lipo` without having Developer Tools installed. The code checks for this anyway.

Tested with a Universal ARM64/x86_64 binary which, using this patch, runs on both now - before, it only runs the last binary in the FAT binary, the other(s) erroring out, as expected.

### CODE REVIEW

**Comment by rschupp** [comment]
> I had to post-process the tmpdir used as par_mktmpdir results in paths with repeated slashes in it which execve doesn't like.

Do you have an example, e.g. what does
```
pp -o par_temp -E 'say $ENV{PAR_TEMP}'
./par_temp
```
print? I can't reproduce this on my Linux box (except when I deliberately set PAR_GLOBAL_TEMP to something that already contains consecutive slashes.

File: myldr/boot.c [hunk 1]
```diff
@@ -77,6 +85,58 @@ int extract_embedded_file(embedded_file_t *emb_file, const char* ext_name, const
...
+/* Simple magic number reader for MacOS */
+#ifdef __MACH__
+uint32_t read_magic(FILE *obj_file, int offset) {
+  uint32_t magic;
+  fseek(obj_file, offset, SEEK_SET);
+  fread(&magic, sizeof(uint32_t), 1, obj_file);
+  return magic;
+}
+
+/* Duplicate argv with some changes */
```
Comment (rschupp) [inline myldr/boot.c:97]: There's no need to create a deep copy of argv, just copy the argv array (including the trailing NULL) and replace argv[0] afterwards. In fact, since we either execve away or die, you might just modify the original argv.

**Comment by plk** [comment]
Example of the double slash from you test program - the execve seems to have trouble with this as I believe that double slash ignoring is mostly a feature of things like shells and higher level programs:

```
/var/folders/9_/96fh404s6mndhsgy4_wcjn3w0000gn/T//par-7068696c6b696d65/cache-c0e725562916c6004b84eccba88a7e9f2218fc76
```

File: myldr/boot.c (hunk 1 above)
Comment (plk) [reply myldr/boot.c:97]: Yes, that is currently a bit of a hack and it also doesn't work as it is currently, looking at it.

File: myldr/boot.c [hunk 2]
```diff
@@ -204,6 +264,66 @@ typedef BOOL (WINAPI *pALLOW)(DWORD);
...
     my_prog = par_findprog(argv[0], par_getenv("PATH"));
+
+#ifdef __MACH__
+    /* Detect if FAT binary */
+    FILE *obj_file = fopen(my_prog, "rb");
+    uint32_t magic = read_magic(obj_file, 0);
+    fclose(obj_file);
+    if (magic == FAT_CIGAM || magic == FAT_MAGIC) {
+
+      /* Create seperate dir for extracted thin binary*/
```
Comment (rschupp) [inline myldr/boot.c:275]: You can't call `par_mktmpdir` a second time (it has global side effects). But at this point you can just use `stmpdir`.

Comment (plk) [reply myldr/boot.c:275]: Ah, yes, I wasn't sure about that. I was originally using `stmpdir` but this seemed not to work as the thin binary, on the second pass in, was being overwritten by the extraction of the custom perl, as mentioned in the code (it extracts to argv[0] name)? The thin binary needs to be argv[0] in the second pass so that the headers are checked.

File: myldr/boot.c [hunk 3]
```diff
@@ -204,6 +264,66 @@ typedef BOOL (WINAPI *pALLOW)(DWORD);
...
+      char *ftmpdir = malloc(strlen(stmpdir)+4);
+      ftmpdir = par_mktmpdir( argv );
+      strcat(ftmpdir, "_FAT");
+      sanitise_tmp(ftmpdir);
+      rc = my_mkdir(ftmpdir, 0700);
```
Comment (rschupp) [inline myldr/boot.c:280]: I prefer to create a subdirectory, e.g. "thin" (w/o arch qualification) under `stempdir` and liposuct^H^H^H^Hextract `my_prog` into that with its original name (you can use `par_basename`). The reason is that boot.c also extracts
the custom perl interpreter into `stmpdir` as `par_basename(my_prog)` and finally execs that. So that when the user runs a packed executable `foo` they will see a `foo` in the process list.

Please use sprintf to create paths, it's hard to follow lotsa strcat's (also your first strcat into archthinbin below may not do what you expect, as malloc doesn't guarantee a zero-filled array).

Comment (plk) [reply myldr/boot.c:280]: Fine, looking at it now.

File: myldr/boot.c (hunk 1 above)
Comment (plk) [reply myldr/boot.c:97]: Resolved this, will update pull request with other bits today. Just changing `argv[0]` now.

File: myldr/boot.c [hunk 4]
```diff
@@ -204,6 +264,66 @@ typedef BOOL (WINAPI *pALLOW)(DWORD);
...
+
+      /* Detect if lipo exists, if not die */
+      struct stat buffer;
+      int exist = stat("/usr/bin/lipo", &buffer);
+      if(exist == 0) {
+        char *archthinbin = malloc(strlen(ftmpdir)+strlen(arch)+11);
+        strcat(archthinbin, ftmpdir);
+        strcat(archthinbin, "/biberthin_");
+        strcat(archthinbin, arch);
+        char *lipocmd = malloc(strlen(arch)+strlen(archthinbin)+strlen(my_prog)+40);
```
Comment (rschupp) [inline myldr/boot.c:300]: Please use fork + execve to run lipo, no need to cobble a command line together, just do
```
char* lipo_argv[] = { "/usr/bin/lipo", "-extract_family", arch, "-output", path, my_prog, NULL }:
```

**Comment by rschupp** [comment]
> Example of the double slash from you test program - the execve seems to have trouble with this as I believe that double slash ignoring is mostly a feature of things like shells and higher level programs

Nope, any *nix I know of ignores multiple slashes on the **syscall** level (except maybe for leading "//"). And my Linux will happily execve stuff with multiple slashes, but maybe MacOS is paranoid just for the exec family...

Anyway, your example is interesting: par_mktmpdir tries the following environment variables:
PAR_TMPDIR, TMPDIR, TEMPDIR, TEMP, TMP. Maybe your environment sets it to something ending in a slash?

**Comment by plk** [comment]
Ah, wasn't aware of that. Yes, it seems this is default for MacOS:

```
[tree][PAR-Packer] > env | fgrep TMP 
TMPDIR=/var/folders/9_/96fh404s6mndhsgy4_wcjn3w0000gn/T/
```

**Comment by plk** [comment]
Should be tidier now as requested and the biber-specific stuff removed.

File: myldr/boot.c (hunk 4 above)
Comment (plk) [reply myldr/boot.c]: Done.

File: myldr/boot.c (hunk 3 above)
Comment (plk) [reply myldr/boot.c:251]: All done

File: myldr/boot.c (hunk 2 above)
Comment (plk) [reply myldr/boot.c:247]: This is addressed in the latest commit
